oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenExchangeAuthenticationProvider.java

```diff
@@ -0,0 +1,313 @@
+		if (StringUtils.hasText(tokenExchangeAuthentication.getActorToken())) {
+			actorAuthorization = this.authorizationService.findByToken(
+					tokenExchangeAuthentication.getActorToken(), OAuth2TokenType.ACCESS_TOKEN);
+			if (actorAuthorization == null) {
+				throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+			}
+
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("Retrieved authorization with actor token");
+			}
+
+			OAuth2Authorization.Token<OAuth2Token> actorToken = actorAuthorization.getToken(
+					tokenExchangeAuthentication.getActorToken());
+			if (!actorToken.isActive()) {
+				// As per https://tools.ietf.org/html/rfc6749#section-5.2
+				// invalid_grant: The provided authorization grant (e.g., authorization code,
+				// resource owner credentials) or refresh token is invalid, expired, revoked [...].
+				throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+			}
+
+			actorPrincipal = actorAuthorization.getAttribute(Principal.class.getName());
+			if (actorPrincipal == null) {
+				// As per https://datatracker.ietf.org/doc/html/rfc8693#section-1.1
+				// We require a principal to be available via the actor_token for
+				// the delegation use case.
+				throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+			}
+		}
+
+		Set<String> authorizedScopes = Collections.emptySet();
+		if (!CollectionUtils.isEmpty(tokenExchangeAuthentication.getScopes())) {
+			for (String requestedScope : tokenExchangeAuthentication.getScopes()) {
+				if (!registeredClient.getScopes().contains(requestedScope)) {
+					throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_SCOPE);
+				}
+			}
+			authorizedScopes = new LinkedHashSet<>(tokenExchangeAuthentication.getScopes());
+		}
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Validated token request parameters");
+		}
+
+		if (CollectionUtils.isEmpty(authorizedScopes)) {
+			// TODO: Validate that original authorizedScopes are also configured for this client?
+			authorizedScopes = new LinkedHashSet<>(subjectAuthorization.getAuthorizedScopes());
+		}
+
+		// @formatter:off
+		DefaultOAuth2TokenContext.Builder tokenContextBuilder = DefaultOAuth2TokenContext.builder()
+				.registeredClient(registeredClient)
+				.authorization(subjectAuthorization)
+				.principal(subjectPrincipal)
+				.authorizationServerContext(AuthorizationServerContextHolder.getContext())
+				.authorizedScopes(authorizedScopes)
+				.tokenType(OAuth2TokenType.ACCESS_TOKEN)
+				.authorizationGrantType(TOKEN_EXCHANGE)
+				.authorizationGrant(tokenExchangeAuthentication);
+		// @formatter:on
+		if (actorPrincipal != null) {
+			// NOTE: Passing attributes requires corresponding changes in JwtGenerator.
+			// TODO: Consider a "composite" principal combining the subject and actor principals?
+			tokenContextBuilder.put(Authentication.class.getName().concat(".ACTOR"), actorPrincipal);
+		}
```

Comment:
I'm trying to work out a way to pass the actor (if any) so it can be used by the `TokenGenerator`. There's a fairly tight coupling between the extra information passed by this `AuthenticationProvider` and the additional claims that should end up in the `access_token`.

There's already a bit of specific code for certain token types in the `JwtGenerator`, but do we want grant type specific functionality in there?

Reply:
`SessionInformation` is passed in via the `OAuth2TokenContext`. Can we do something similar?